### PR TITLE
DROOLS-4118: [DMN Designer] Documentation tab - Show information from the DMN diagram in the documentation tab

### PR DESCRIPTION
**Pull request by karreiro**
See:
- https://issues.jboss.org/browse/DROOLS-4117
- https://issues.jboss.org/browse/DROOLS-4118

---

**Demo:**

---

@tiagodolphine I've applied the use of  `PathHelper` on stunner (`BPMNDocumentationView.java`). I just tried the **Documentation** tab there and everything looks fine, but I'm adding you for double-checking it ;-) (thus, feel free to just review stunner related classes):

---

Part of an ensemble:
- https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/971
- https://github.com/kiegroup/kie-wb-common/pull/2694

### CODE REVIEW

**Comment by karreiro** [comment]
Jenkins, please retest this.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationService.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.kie.workbench.common.dmn.client.editors.documentation.common;
+
+import org.kie.workbench.common.stunner.core.diagram.Diagram;
+import org.kie.workbench.common.stunner.core.documentation.model.DocumentationOutput;
+import org.kie.workbench.common.stunner.core.documentation.model.HTMLDocumentationTemplate;
+import org.kie.workbench.common.stunner.core.documentation.service.DiagramDocumentationService;
+
+public interface DMNDocumentationService extends DiagramDocumentationService<
+        Diagram,
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationService.java]: Thank you, Mike :-) Fixed!

**Comment by jomarko** [comment]
### Manual Review - Note#01
The selected nodes are highlighted also in documentation, should I report this as an issue?

Screenshot from 2019-06-13 11-31-38

**Comment by jomarko** [comment]
### Manaul Review - Note#02
Background colors are not in documentation. Changed Font is.
Screenshot from 2019-06-13 11-34-24

Screenshot from 2019-06-13 11-34-29

**Comment by jomarko** [comment]
### Manual Review - Note#03
The print button looks different for BPMN and DMN

DMN
Screenshot from 2019-06-13 11-36-04

BPMN
Screenshot from 2019-06-13 11-36-11

**Comment by jomarko** [comment]
Manual review finished

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                                       hasGraphNodes(diagram),
+                                       getDataTypes(diagram),
+                                       getDiagramImage(),
+                                       getCurrentUserName(),
+                                       getCurrentDate(),
+                                       getDocumentationI18n());
+    }
+
+    String getFileName(final Diagram diagram) {
+        Metadata metadata = diagram.getMetadata();
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: can be also `final`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        dataTypes.add(dataType);
+
+        makeDMNDocumentationDataTypes(itemDefinition.getItemComponent(), dataTypes, level + 1);
+    }
+
+    private String getConstraint(final ItemDefinition itemDefinition) {
+        final UnaryTests allowedValues = itemDefinition.getAllowedValues();
+
+        if (allowedValues != null && !allowedValues.getText().getValue().isEmpty()) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: Could we please use `StringUtils` ?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                .getDefinitions(diagram)
+                .getDescription()
+                .getValue();
+    }
+
+    private String getCurrentUserName() {
+        return sessionInfo.getIdentity().getIdentifier();
+    }
+
+    private Optional<AbstractCanvasHandler> getCanvasHandler() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: I am not sure if this isn't already present in `DMNGraphUtils.getCanvasHandler`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties [hunk 5]
```diff
@@ -245,3 +245,18 @@ DecisionComponentsView.EmptyStateDescription=To add an external components to th
...
+DMNDocumentationI18n.GeneratedFrom=Generated from
+DMNDocumentationI18n.TableOfContents=Table of Contents
+DMNDocumentationI18n.DmnModel=DMN model
+DMNDocumentationI18n.DataTypes=Data Types
+DMNDocumentationI18n.DrdComponents=DRD components
+DMNDocumentationI18n.NoDataTypes=No data types.
+DMNDocumentationI18n.DiagramDoesNotHaveElements=The diagram does not have any element.
+DMNDocumentationFactory.NoFileName=No file name
+DMNDocumentationFactory.Constraints=Constraints :
+DMNDocumentationFactory.ListYes=List: Yes
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: Slightly inconsistency I think:
```
Constraints :
List : yes
```

vs

```
Constraints:
List: yes
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java (hunk 2 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: Fixed! Thanks :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java (hunk 3 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: Fixed! Thanks :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java (hunk 4 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java]: Fixed! Thanks :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties (hunk 5 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: Fixed! Thanks :-)

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/documentation/BPMNDocumentationView.java [hunk 6]
```diff
@@ -92,61 +99,8 @@ public BPMNDocumentationView initialize(Diagram diagram) {
...
-            }, 10);
-        };
-
-        return true;
-    }-*/;
-
-    private void print() {
-        print(documentationDiv);
+    void print() {
+        printHelper.print(documentationDiv);
```
Comment (tiagodolphine) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/documentation/BPMNDocumentationView.java]: It looks good to me the use of the printHelper here, in this way the print logic in centralized there using elmental2, good change. Thanks.

Comment (karreiro) [reply kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/documentation/BPMNDocumentationView.java]: Thank you, @tiagodolphine ☺️

**Comment by karreiro** [comment]
Thank you for the review, @jomarko. Changes applied.

---

> **Manual Review - Note#01**
> The selected nodes are highlighted also in documentation, should I report this as an issue?

Since this behaviour happens in the Stunner editor, and every place the exports a canvas, I think that we can handle it in a separated JIRA.


> **Manaul Review - Note#02**
> Background colors are not in documentation. Changed Font is.

Fixed. Good catch!




> **Manual Review - Note#03**
> The print button looks different for BPMN and DMN

I don't think this is a problem. I've checked the BC and seems like a standard to use the icon in the left side, see:
Screen Shot 2019-06-13 at 11 09 33
Screen Shot 2019-06-13 at 11 07 00
Screen Shot 2019-06-13 at 11 06 42
Screen Shot 2019-06-13 at 11 05 09

Thus, I think we can keep as is for now, and check in the Monday meeting if the button is fine like this. If not, we can fix it in a separated JIRA.

Thanks again for the review :-)

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
The issue for **Manual Review Note#01** reported here https://issues.jboss.org/browse/DROOLS-4178

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this 😥

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by karreiro** [comment]
(rebasing the branch...) 👀

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
(fixing conflicts...) 👀

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
@tiagodolphine Moving `PrintHelper` from (widgets):
`org.kie.workbench.common.widgets.client.util.PrintHelper;`
to (stunner):
`org.kie.workbench.common.stunner.core.client.util.PrintHelper;`

**Comment by tiagodolphine** [comment]
Ok @karreiro LGTM
Thanks.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Pull request merged**